rllib/core/models/base.py

```diff
@@ -0,0 +1,380 @@
+                }
+
+        @dataclass
+        class NumpyEncoderConfig(ModelConfig):
+            factor: int = None
+
+            def build(self, framework: str):
+                return NumpyEncoder(self)
+
+        config = NumpyEncoderConfig(factor=2)
+        encoder = NumpyEncoder(config)
+        print(encoder({SampleBatch.OBS: 1, STATE_IN: 2}))
+
+    .. testoutput::
+
+        {'encoder_out': 2, 'state_out': 4}
+
+    """
+
+    @override(Model)
+    def get_input_spec(self) -> Union[Spec, None]:
+        return convert_to_canonical_format([SampleBatch.OBS, STATE_IN])
+
+    @override(Model)
+    def get_output_spec(self) -> Union[Spec, None]:
+        return convert_to_canonical_format([ENCODER_OUT, STATE_OUT])
+
+    @abc.abstractmethod
+    def _forward(self, input_dict: NestedDict, **kwargs) -> NestedDict:
+        """Returns the latent of the encoder for the given inputs.
+
+        This method is called by the forwarding method of the respective framework
+        that is itself wrapped by RLlib in order to check model inputs and outputs.
+
+        The input dict contains at minimum the observation and the state of the encoder.
+        The output dict contains at minimum the latent and the state of the encoder.
+        These values have the keys `SampleBatch.OBS` and `STATE_IN` in the inputs, and
+        `STATE_OUT` and `ENCODER_OUT` and outputs to establish an agreement
+        between the encoder and RLModules. For stateless encoders, states can be None.
+
+        Args:
+            input_dict: The input tensors.
+            **kwargs: Forward compatibility kwargs.
+
+        Returns:
+            NestedDict: The output tensors.
+        """
+        raise NotImplementedError
+
+
+class ActorCriticEncoder(Encoder):
+    """An encoder that potentially holds two encoders.
+
+    This is a special case of encoder that can either enclose a single,
+    shared encoder or two separate encoders: One for the actor and one for the
+    critic. The two encoders are of the same type and we can therefore make the
+    assumption that they have the same input and output specs.
+    """
+
+    framework = None
+
+    def __init__(self, config: ModelConfig) -> None:
+        if config.shared:
+            self.encoder = config.base_encoder_config.build(framework=self.framework)
+        else:
+            self.actor_encoder = config.base_encoder_config.build(
+                framework=self.framework
+            )
+            self.critic_encoder = config.base_encoder_config.build(
+                framework=self.framework
+            )
+
+        # We need to call Encoder.__init__() after initializing the encoder(s) in
+        # order to build on their specs.
+        super().__init__(config)
+
+    @override(Model)
+    def get_input_spec(self) -> Union[Spec, None]:
+        if self.config.shared:
+            state_in_spec = self.encoder.input_spec[STATE_IN]
+        else:
+            state_in_spec = NestedDict(
```

Comment:
nit: to make it easier to read this can just be a normal dict. doesn't have to be NestedDict